// @sanity/form-builder/lib/inputs/arrays/ArrayOfObjectsInput/ArrayInput.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.ArrayInput = void 0;

var _isPlainObject2 = _interopRequireDefault(require("lodash/isPlainObject"));

var _types = require("@sanity/types");

var _paths = require("@sanity/util/paths");

var _components = require("@sanity/base/components");

var _ui = require("@sanity/ui");

var _react = _interopRequireDefault(require("react"));

var _operators = require("rxjs/operators");

var _legacyParts = require("../../../legacyParts");

var _PatchEvent = require("../../../PatchEvent");

var _resolveTypeName = require("../../../utils/resolveTypeName");

var _Alert = require("../../../components/Alert");

var _Details = require("../../../components/Details");

var _randomKey = _interopRequireDefault(require("../common/randomKey"));

var _list = require("../common/list");

var _item = require("./item");

var _uploadTarget = require("./uploadTarget/uploadTarget");

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

function ownKeys(object, enumerableOnly) { var keys = Object.keys(object); if (Object.getOwnPropertySymbols) { var symbols = Object.getOwnPropertySymbols(object); if (enumerableOnly) symbols = symbols.filter(function (sym) { return Object.getOwnPropertyDescriptor(object, sym).enumerable; }); keys.push.apply(keys, symbols); } return keys; }

function _objectSpread(target) { for (var i = 1; i < arguments.length; i++) { var source = arguments[i] != null ? arguments[i] : {}; if (i % 2) { ownKeys(Object(source), true).forEach(function (key) { _defineProperty(target, key, source[key]); }); } else if (Object.getOwnPropertyDescriptors) { Object.defineProperties(target, Object.getOwnPropertyDescriptors(source)); } else { ownKeys(Object(source)).forEach(function (key) { Object.defineProperty(target, key, Object.getOwnPropertyDescriptor(source, key)); }); } } return target; }

function _defineProperty(obj, key, value) { if (key in obj) { Object.defineProperty(obj, key, { value: value, enumerable: true, configurable: true, writable: true }); } else { obj[key] = value; } return obj; }

var UploadTargetFieldset = (0, _uploadTarget.uploadTarget)(_components.FormFieldSet);
var NO_MARKERS = [];

function createProtoValue(type) {
  if (!(0, _types.isObjectSchemaType)(type)) {
    throw new Error("Invalid item type: \"".concat(type.type, "\". Default array input can only contain objects (for now)"));
  }

  var _key = (0, _randomKey.default)(12);

  return type.name === 'object' ? {
    _key
  } : {
    _type: type.name,
    _key
  };
}

class ArrayInput extends _react.default.Component {
  constructor() {
    super(...arguments);

    _defineProperty(this, "_focusArea", null);

    _defineProperty(this, "uploadSubscriptions", {});

    _defineProperty(this, "insert", (itemValue, position, atIndex) => {
      var onChange = this.props.onChange;
      onChange(_PatchEvent.PatchEvent.from((0, _PatchEvent.setIfMissing)([]), (0, _PatchEvent.insert)([itemValue], position, [atIndex])));
    });

    _defineProperty(this, "handlePrepend", value => {
      this.insert(value, 'before', 0);
      this.handleFocusItem(value);
    });

    _defineProperty(this, "handleAppend", value => {
      this.insert(value, 'after', -1);
      this.handleFocusItem(value);
    });

    _defineProperty(this, "handleRemoveItem", item => {
      this.removeItem(item);
    });

    _defineProperty(this, "handleFocus", () => {
      this.props.onFocus([]);
    });

    _defineProperty(this, "handleFocusItem", item => {
      this.props.onFocus([{
        _key: item._key
      }, _paths.FOCUS_TERMINATOR]);
    });

    _defineProperty(this, "handleItemChange", (event, item) => {
      var _this$props = this.props,
          onChange = _this$props.onChange,
          value = _this$props.value;
      var memberType = this.getMemberTypeOfItem(item);

      if (!memberType) {
        // eslint-disable-next-line no-console
        console.log('Could not find member type of item ', item);
        return;
      }

      if (memberType.readOnly) {
        return;
      }

      var key = item._key || (0, _randomKey.default)(12);
      onChange(event.prefixAll({
        _key: key
      }).prepend(item._key ? [] : (0, _PatchEvent.set)(key, [value.indexOf(item), '_key'])));
    });

    _defineProperty(this, "handleSortEnd", event => {
      var _this$props2 = this.props,
          value = _this$props2.value,
          onChange = _this$props2.onChange;
      var item = value[event.oldIndex];
      var refItem = value[event.newIndex];

      if (!item._key || !refItem._key) {
        // eslint-disable-next-line no-console
        console.error('Neither the item you are moving nor the item you are moving to have a key. Cannot continue.');
        return;
      }

      if (event.oldIndex === event.newIndex || item._key === refItem._key) {
        return;
      }

      onChange(_PatchEvent.PatchEvent.from((0, _PatchEvent.unset)([{
        _key: item._key
      }]), (0, _PatchEvent.insert)([item], event.oldIndex > event.newIndex ? 'before' : 'after', [{
        _key: refItem._key
      }])));
    });

    _defineProperty(this, "setFocusArea", el => {
      this._focusArea = el;
    });

    _defineProperty(this, "getUploadOptions", file => {
      var _this$props3 = this.props,
          type = _this$props3.type,
          resolveUploader = _this$props3.resolveUploader;

      if (!resolveUploader) {
        return [];
      }

      return type.of.map(memberType => ({
        type: memberType,
        uploader: resolveUploader(memberType, file)
      })).filter(member => member.uploader);
    });

    _defineProperty(this, "handleFixMissingKeys", () => {
      var _this$props4 = this.props,
          onChange = _this$props4.onChange,
          value = _this$props4.value;
      var patches = value.map((val, i) => (0, _PatchEvent.setIfMissing)((0, _randomKey.default)(), [i, '_key']));
      onChange(_PatchEvent.PatchEvent.from(...patches));
    });

    _defineProperty(this, "handleRemoveNonObjectValues", () => {
      var _this$props5 = this.props,
          onChange = _this$props5.onChange,
          value = _this$props5.value;
      var nonObjects = value.reduce((acc, val, i) => (0, _isPlainObject2.default)(val) ? acc : acc.concat(i), []).reverse();
      var patches = nonObjects.map(index => (0, _PatchEvent.unset)([index]));
      onChange(_PatchEvent.PatchEvent.from(...patches));
    });

    _defineProperty(this, "handleUpload", (_ref) => {
      var file = _ref.file,
          type = _ref.type,
          uploader = _ref.uploader;
      var onChange = this.props.onChange;
      var item = createProtoValue(type);
      var key = item._key;
      this.insert(item, 'after', -1);
      var events$ = uploader.upload(file, type).pipe((0, _operators.map)(uploadEvent => _PatchEvent.PatchEvent.from(uploadEvent.patches || []).prefixAll({
        _key: key
      })));
      this.uploadSubscriptions = _objectSpread(_objectSpread({}, this.uploadSubscriptions), {}, {
        [key]: events$.subscribe(onChange)
      });
    });
  }

  removeItem(item) {
    var _this$props6 = this.props,
        onChange = _this$props6.onChange,
        onFocus = _this$props6.onFocus,
        value = _this$props6.value;
    onChange(_PatchEvent.PatchEvent.from((0, _PatchEvent.unset)(item._key ? [{
      _key: item._key
    }] : [value.indexOf(item)])));

    if (item._key in this.uploadSubscriptions) {
      this.uploadSubscriptions[item._key].unsubscribe();
    }

    var idx = value.indexOf(item);
    var nextItem = value[idx + 1] || value[idx - 1];
    onFocus([nextItem ? {
      _key: nextItem._key
    } : _paths.FOCUS_TERMINATOR]);
  }

  getMemberTypeOfItem(item) {
    var type = this.props.type;
    var itemTypeName = (0, _resolveTypeName.resolveTypeName)(item);
    return type.of.find(memberType => memberType.name === itemTypeName);
  }

  focus() {
    if (this._focusArea) {
      this._focusArea.focus();
    }
  }

  render() {
    var _this$props7 = this.props,
        type = _this$props7.type,
        _this$props7$level = _this$props7.level,
        level = _this$props7$level === void 0 ? 1 : _this$props7$level,
        markers = _this$props7.markers,
        readOnly = _this$props7.readOnly,
        onChange = _this$props7.onChange,
        _this$props7$value = _this$props7.value,
        value = _this$props7$value === void 0 ? [] : _this$props7$value,
        presence = _this$props7.presence,
        focusPath = _this$props7.focusPath,
        onBlur = _this$props7.onBlur,
        onFocus = _this$props7.onFocus,
        compareValue = _this$props7.compareValue,
        filterField = _this$props7.filterField,
        directUploads = _this$props7.directUploads;
    var hasNonObjectValues = (value || []).some(item => !(0, _isPlainObject2.default)(item));

    if (hasNonObjectValues) {
      return /*#__PURE__*/_react.default.createElement(_components.FormFieldSet, {
        title: type.title,
        description: type.description,
        level: level - 1,
        tabIndex: 0,
        onFocus: this.handleFocus,
        ref: this.setFocusArea,
        __unstable_markers: markers
      }, /*#__PURE__*/_react.default.createElement(_Alert.Alert, {
        status: "error",
        suffix: /*#__PURE__*/_react.default.createElement(_ui.Stack, {
          padding: 2
        }, /*#__PURE__*/_react.default.createElement(_ui.Button, {
          onClick: this.handleRemoveNonObjectValues,
          text: "Remove non-object values",
          tone: "critical"
        })),
        title: /*#__PURE__*/_react.default.createElement(_react.default.Fragment, null, "Invalid list values")
      }, /*#__PURE__*/_react.default.createElement(_ui.Text, {
        as: "p",
        muted: true,
        size: 1
      }, "Some items in this list are not objects. This must be fixed in order to edit the list."), /*#__PURE__*/_react.default.createElement(_Details.Details, {
        marginTop: 4,
        open: __DEV__,
        title: /*#__PURE__*/_react.default.createElement(_react.default.Fragment, null, "Developer info")
      }, /*#__PURE__*/_react.default.createElement(_ui.Stack, {
        space: 3
      }, /*#__PURE__*/_react.default.createElement(_ui.Text, {
        as: "p",
        muted: true,
        size: 1
      }, "This usually happens when items are created using an API client, or when a custom input component has added invalid data to the list.")))));
    }

    var options = type.options || {};
    var hasMissingKeys = value.some(item => !item._key);
    var isSortable = options.sortable !== false && !hasMissingKeys;
    var isGrid = options.layout === 'grid';
    return /*#__PURE__*/_react.default.createElement(UploadTargetFieldset, {
      __unstable_changeIndicator: false,
      tabIndex: 0,
      title: type.title,
      description: type.description,
      onFocus: this.handleFocus,
      onBlur: onBlur,
      level: level - 1,
      __unstable_presence: presence.filter(item => item.path[0] === '$'),
      __unstable_markers: markers,
      disabled: !directUploads || readOnly,
      ref: this.setFocusArea,
      getUploadOptions: this.getUploadOptions,
      onUpload: this.handleUpload
    }, /*#__PURE__*/_react.default.createElement(_ui.Stack, {
      space: 3
    }, hasMissingKeys && /*#__PURE__*/_react.default.createElement(_Alert.Alert, {
      status: "warning",
      suffix: /*#__PURE__*/_react.default.createElement(_ui.Stack, {
        padding: 2
      }, /*#__PURE__*/_react.default.createElement(_ui.Button, {
        onClick: this.handleFixMissingKeys,
        text: "Add missing keys",
        tone: "caution"
      })),
      title: /*#__PURE__*/_react.default.createElement(_react.default.Fragment, null, "Missing keys")
    }, /*#__PURE__*/_react.default.createElement(_ui.Text, {
      as: "p",
      muted: true,
      size: 1
    }, "Some items in the list are missing their keys. This must be fixed in order to edit the list."), /*#__PURE__*/_react.default.createElement(_Details.Details, {
      marginTop: 4,
      open: __DEV__,
      title: /*#__PURE__*/_react.default.createElement(_react.default.Fragment, null, "Developer info")
    }, /*#__PURE__*/_react.default.createElement(_ui.Stack, {
      space: 3
    }, /*#__PURE__*/_react.default.createElement(_ui.Text, {
      as: "p",
      muted: true,
      size: 1
    }, "This usually happens when items are created using an API client, and the", ' ', /*#__PURE__*/_react.default.createElement("code", null, "_key"), " property has not been included."), /*#__PURE__*/_react.default.createElement(_ui.Text, {
      as: "p",
      muted: true,
      size: 1
    }, "The value of the ", /*#__PURE__*/_react.default.createElement("code", null, "_key"), " property must be a unique string.")))), /*#__PURE__*/_react.default.createElement(_ui.Stack, {
      "data-ui": "ArrayInput__content",
      space: 3
    }, value && value.length > 0 && /*#__PURE__*/_react.default.createElement(_list.List, {
      onSortEnd: this.handleSortEnd,
      isSortable: isSortable,
      isGrid: isGrid
    }, value.map((item, index) => {
      var isChildMarker = marker => (0, _paths.startsWith)([index], marker.path) || (0, _paths.startsWith)([{
        _key: item && item._key
      }], marker.path);

      var childMarkers = markers.filter(isChildMarker);

      var isChildPresence = pItem => (0, _paths.startsWith)([index], pItem.path) || (0, _paths.startsWith)([{
        _key: item && item._key
      }], pItem.path);

      var childPresence = presence.filter(isChildPresence);
      return /*#__PURE__*/_react.default.createElement(_list.Item, {
        key: item._key || index,
        isSortable: isSortable,
        isGrid: isGrid,
        index: index
      }, /*#__PURE__*/_react.default.createElement(_item.ArrayItem, {
        compareValue: compareValue,
        filterField: filterField,
        focusPath: focusPath,
        index: index,
        markers: childMarkers.length === 0 ? NO_MARKERS : childMarkers,
        onBlur: onBlur,
        onChange: this.handleItemChange,
        onFocus: onFocus,
        onRemove: this.handleRemoveItem,
        presence: childPresence,
        readOnly: readOnly || hasMissingKeys,
        type: type,
        value: item
      }));
    })), /*#__PURE__*/_react.default.createElement(_legacyParts.ArrayFunctions, {
      type: type,
      value: value,
      readOnly: readOnly,
      onAppendItem: this.handleAppend,
      onPrependItem: this.handlePrepend,
      onFocusItem: onFocus,
      onCreateValue: createProtoValue,
      onChange: onChange
    }))));
  }

}

exports.ArrayInput = ArrayInput;

_defineProperty(ArrayInput, "defaultProps", {
  focusPath: []
});
